src/api.js

```diff
@@ -0,0 +1,187 @@
+const http = require('http');
+const url = require('url');
+const express = require('express');
+const router = express.Router();
+const fs = require('fs');
+const config = require('./config');
+const tableMap = require('./tableMap');
+const {
+  MissingArgumentError,
+  MalformedArgumentError,
+  BigQueryManager,
+  MemcachedManager,
+} = require('./helperClasses');
+const cacheManager = new MemcachedManager('127.0.0.1:11211');
+
+/**
+* returns the table containing records from the given app id
+* @param{string} id the app id
+* @returns{string} the table name in SQL table format
+*/
+const getTable = function (map, id) {
+  const regex = /(^[a-z]{2,3})([\.][a-z0-9]*)+$/gmi;
+  if (!id.match(regex)) {
+    const msg = `cannot parse app id: '${id}'. Check formatting and try again`;
+    throw new MalformedArgumentError(msg);
+  }
+  const obj = map[id];
+  if (!obj ) {
+    const msg = `could not find a table for app id: '${id}'`;
+    const err = new Error(msg);
+    throw err;
+  }
+  return `${obj.project}.${obj.dataset}.${obj.table}`;
+}
+
+function sendRows(res, rows, nextCursor, key) {
+  if (!rows || rows.length == 0 || !nextCursor) {
+    cacheManager.removeCache(key);
+  }
+  const resObj = formatRowsToJson(rows);
+  return res.status(200).json({data: resObj, nextCursor});
+}
+
+function formatRowsToJson (rows) {
+  let resObj = [];
+  if (!rows || rows.length === 0) return resObj;
+  rows.forEach((row) => {
+    resObj.push({
+      attribution_url: row.referral_source,
+      app_id: row.app_package_name,
+      ordered_id: row.event_timestamp,
+      user: {
+        id: row.user_pseudo_id,
+        metadata: {
+          continent: row.continent,
+          country: row.country,
+          region: row.region,
+          city: row.city,
+        },
+        ad_attribution: {
+          source: getSource(row.referral_source),
+          data: {
+            advertising_id: row.advertising_id,
+          },
+        },
+        event: {
+          name: row.event_name,
+          date: row.event_date,
+          timestamp: row.event_timestamp,
+          label: row.label,
+          action: row.action,
+          value: row.value,
+          value_type: row.type,
+        }
+      }
+
+    });
+  });
+  return resObj;
+}
+
+function getSource(referralString) {
+  if (!referralString) return "no-source";
+  const regex = /^([a-z]{1,6})(_[a-z]{1,3})?/gmi;
+  const group = referralString.match(regex);
+  switch (group.toString().toLowerCase()) {
+    case 'fb_web':
```

Comment:
This looks like data, I would bring it into config (a json object mapping name to name).

Reply:
ooh, good point